pkg/predicate/has_finalizer.go

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package predicate
+
+import (
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// HasFinalizer returns a predicate that detects if the object has the given finalizer
+// This is used to not requeue all secrets in the cluster (which might be quite a lot),
+// but only requeue secrets from create/update events with the controller's finalizer.
+// This is to ensure, that we properly remove the finalizer in case we missed an important
+// update event for a ManagedResource.
+func HasFinalizer(finalizer string) predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			// Create event is emitted on start-up, when the cache is populated from a complete list call for the first time.
+			// We should enqueue all secrets, which have the controller's finalizer in order to remove it in case we missed
+			// an important update event to a ManagedResource during downtime.
+			if e.Meta == nil {
+				log.Error(nil, "Create event has no object meta", "event", e)
+				return false
+			}
+
+			return sets.NewString(e.Meta.GetFinalizers()...).Has(finalizer)
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			// We only need to check MetaNew. If the finalizer was in MetaOld and is not in MetaNew, it is already
+			// removed and we don't need to reconcile the secret.
+			if e.MetaNew == nil {
+				log.Error(nil, "Update event has no new object meta", "event", e)
+				return false
+			}
+
+			return sets.NewString(e.MetaNew.GetFinalizers()...).Has(finalizer)
```

Comment:
Can this be put into a separate function that you can call in both `CreateFunc` and `UpdateFunc`?